### PR TITLE
Fix coils alignment

### LINKED ISSUE #18520: [BUG] Coils alignment after home
@Fabi0San 

I suspect there is an error in code to align steppers on coils after home.

In motion.cpp, function `backout_to_tmc_homing_phase` is always used `Z_HOME_DIR`.

`const float mmDelta = -(int16_t(phaseDelta / axisMicrostepSize) * planner.steps_to_mm[axis] * (Z_HOME_DIR));`

I'm not sure if the fix is only to consider X/Y/Z_HOME_DIR in line above or replace `Z_HOME_DIR` with `invertDir` and add an `XOR` to `invertDir` to consider home direction

eg `invertDir = INVERT_X_DIR ^ (X_HOME_DIR==-1);`

### PR DESCRIPTION
**Pull request by GMagician**
This should fix coils alignment in all cases

Should fix #18520 
@Fabi0San please validate this fix

### CODE REVIEW

**Comment by GMagician** [comment]
OT: Is there any possibility for people with no write access to enable labels?

**Comment by thisiskeithb** [comment]
> OT: Is there any possibility for people with no write access to enable labels?

Github's permissions are not very granular, but I added some tags.

**Comment by GMagician** [comment]
> Github's permissions are not very granular, but I added some tags.

The possibility to add them will help a lot, all of you, to immediately identify what is done, but it's ok

**Comment by thinkyhead** [comment]
It looks correct to me. But, since I didn't know enough to see the problem in the original homing phase code, I may not be the best judge of the new code either.

**Comment by GMagician** [comment]
@thinkyhead I really didn't had a fault to fix it. I was just having a Z issue (not repeatable position) and I saw the message (in debug mode). I checked code and I figured out it was not correct, this is why I asked @Fabi0San to take a look at it

**Comment by Fabi0San** [comment]
@GMagician you are absolutely right, with my delta head I made the assumption that Z home direction is the only one that is.

This would have been easy to spot if M122 showed the stepper phase, I have that change on a branch somewhere, I'll add it to this PR and test everything a bit later today or tomorrow. 
It'll be easier if you give me push perms on your branch, Other wise I'll send in a pr to it for you to merge.

**Comment by GMagician** [comment]
@Fabi0San not sure it's possible to give write access to a branch, if it's possible, please guide me to such procedure..thanks

**Comment by thinkyhead** [comment]
> not sure it's possible to give write access to a branch

A GitHub user can submit a PR to any fork + branch that they check out, modify, and push to up to their own fork (`origin`).

If a PR is submitted by Fabi0San to the `GMagician` branch "`Fix-coils-alignment`" then as soon as you merge that PR the new commits will be added to this PR — just as if you added the commits to your `Fix-coils-alignment` branch by any other method.

**Comment by Fabi0San** [comment]
So I just had a chance to look at it. The issue is real, not only I assumed that every axis homed in the Z direction but also made some assumptions that would work only for home to max  direction setups.

That said, @GMagician's fix doesn't quite address it. It is effectively just flipping direction on everything homing to max, which includes all the deltas and may cause some endstop crashes.

I'll send out my suggested changes in a bit.

**Comment by Fabi0San** [comment]
>That said, @GMagician's fix doesn't quite address it. It is effectively just flipping direction on everything homing to max, which includes all the deltas and may cause some endstop crashes.

I stand corrected, @GMagician's fix was mostly effective with exception of the removed int rounding, but that code is confusing as hell so I cleaned it all up, gave variables better names and also added lots of comments.

I also included my changes to M122 which exposes the phase counter and other very useful registers for tuning and debugging TMC drivers.

I created PR https://github.com/GMagician/Marlin/pull/1, please merge it.

>A GitHub user can submit a PR to any fork + branch

Alternatively @GMagician can temporarily grant me contributor status to his Marlin repo.

I'll run a few tests tomorrow on my delta, will report back.

**Comment by GMagician** [comment]
@Fabi0San you have access to repository

File: Marlin/src/module/motion.cpp [hunk 1]
```diff
@@ -1502,8 +1502,8 @@ void set_axis_not_trusted(const AxisEnum axis) {
     // Skip to next if target position is behind current. So it only moves away from endstop.
     if (phaseDelta < 0) phaseDelta += 1024;
 
-    // Convert tmc µsteps(phase) to whole Marlin µsteps to mm to effector backout direction
-    const float mmDelta = int16_t(phaseDelta / phasePerUStep) * planner.steps_to_mm[axis] * effectorBackoutDir ;
+    // Convert tmc µsteps(phase) to whole Marlin µsteps to effector to backout direction to mm 
+    const float mmDelta = int16_t(phaseDelta / phasePerUStep) * effectorBackoutDir * planner.steps_to_mm[axis];
```
Comment (GMagician) [inline Marlin/src/module/motion.cpp:1506]: Don't you lose precision dividing integers?

Comment (Fabi0San) [reply Marlin/src/module/motion.cpp:1506]: It is very very much intentional, not every phase is addressable, we want to consistently stop before we overflow into the next marlin microstep. 
For instance on a system running 16 marlin microsteps per stepper full step advances 16 tmc microsteps per marlin microstep, all good there but the driver is initialized on phase 8, so with a single marlin microstep you get to phase 24, making say phase 16 unreachable.

**Comment by Fabi0San** [comment]
I'm still testing this btw so hold the merge for now.

File: Marlin/src/module/motion.cpp [hunk 2]
```diff
@@ -1468,23 +1468,23 @@ void set_axis_not_trusted(const AxisEnum axis) {
...
+          stepperBackoutDir = INVERT_Y_DIR ? effectorBackoutDir : -effectorBackoutDir;
           break;
       #endif
       #ifdef Z_MICROSTEPS
         case Z_AXIS:
           phasePerUStep = 256 / (Z_MICROSTEPS);
           phaseCurrent = stepperZ.get_microstep_counter();
           effectorBackoutDir = -Z_HOME_DIR;
-          stepperBackoutDir = INVERT_Y_DIR ? -effectorBackoutDir : effectorBackoutDir;
+          stepperBackoutDir = INVERT_Z_DIR ? effectorBackoutDir : -effectorBackoutDir;
```
Comment (GMagician) [inline Marlin/src/module/motion.cpp:1487]: Sign doesn't seems correct now.
Lets say invert = false and home dir -1,phase 128, stopped at 126.
you need to move 2 usteps positive to reach next phase but you consider 1022 usteps
Am I wrong?

Comment (GMagician) [inline Marlin/src/module/motion.cpp:1487]: I think sign is not correct now.
Let's say invert=false, home dir = -1, phase =128 and stop pos =126
You need to move positive 2 ustep but on your computation you get `effectorBackoutDir=1` and `stepperBackoutDir=-1` then phaseDelta is -2 you will move 1022 positive

**Comment by Fabi0San** [comment]
I've tested this and it works properly on my machine where all axes home the same direction but not all steppers are mounted the same direction. 

```
#define INVERT_X_DIR false
#define INVERT_Y_DIR true
#define INVERT_Z_DIR true
#define X_HOME_DIR 1  // deltas always home to max
#define Y_HOME_DIR 1
#define Z_HOME_DIR 1
#define TMC_HOME_PHASE { 896, 896, 896 }
```
```
M111 S38 ; enable debug msg
M666 X0 Y0 Z0 ;disable end stop trim on deltas
G28
>>> G28
Machine Type: Delta
Endstop X hit at Phase:452 Delta:444 Distance:-0.13736
Endstop Y hit at Phase:692 Delta:820 Distance:-0.25500
Endstop Z hit at Phase:940 Delta:44 Distance:-0.01249
<<< G28
M122
uStep count     892     900     900 
```
896 is unreachable so steppers stay consistently close to it depending which side they come from.

@GMagician can you please test these changes on your machine?

File: Marlin/src/module/motion.cpp (hunk 2 above)
Comment (Fabi0San) [reply Marlin/src/module/motion.cpp:1487]: The key on all this is the relationship between the marlin coordinate increasing on an non inverted stepper and the tmc phase increasing. They seem to be opposite to each other. I hope it is consistent across all TMC drivers. I'll read some data sheets now on the direction pin definition.

**Comment by GMagician** [comment]
That doesn't seems correct either..let's consider X
invert = false home = +1 you stop at 452 with desired phase at 896.
You can't go to next positive phase so you have to move back to previous then 1024-896+452=580 usteps then you have to move back (-580*usteptommfactor). I'm just thinking how it should works, not read any document

File: Marlin/src/module/motion.cpp (hunk 2 above)
Comment (Fabi0San) [reply Marlin/src/module/motion.cpp:1487]: From the 2209 datasheet:
> the microstep counter is increased (DIR=0) or decreased (DIR=1) with each STEP pulse

now we need to check what is the DIR state when marlin increases the axis coordinates, or maybe @thinkyhead knows from the top of his head.

**Comment by GMagician** [comment]
> @GMagician can you please test these changes on your machine?

At the moment I can't I just discovered that my printer has Z not 90° (I'm building it but was all mounted) so I'm going to dismount all component to properly align (it's a bear)

File: Marlin/src/module/motion.cpp (hunk 2 above)
Comment (GMagician) [reply Marlin/src/module/motion.cpp:1487]: Yeah, I supposed invert=false microsteps increment but may not be true

Comment (GMagician) [reply Marlin/src/module/motion.cpp:1487]:
```c++
  #define SET_STEP_DIR(A)                       \
    if (motor_direction(_AXIS(A))) {            \
      A##_APPLY_DIR(INVERT_##A##_DIR, false);   \
      count_direction[_AXIS(A)] = -1;           \
    }                                           \
    else {                                      \
      A##_APPLY_DIR(!INVERT_##A##_DIR, false);  \
      count_direction[_AXIS(A)] = 1;            \
    }
```

with 
`#define X_APPLY_DIR(v,Q)  X_DIR_WRITE(v)`
`#define X_DIR_WRITE(STATE) WRITE(X_DIR_PIN,STATE)`

Then it seems negative count is with dir=0. My assumption was wrong

Comment (GMagician) [reply Marlin/src/module/motion.cpp:1487]: I think it's better to add a comment on code, just to highlight this

**Comment by Fabi0San** [comment]
> That doesn't seems correct either..let's consider X

Though it is landing on the right spot, but we surely should understand why.

> invert = false home = +1 you stop at 452 with desired phase at 896.
> You can't go to next positive phase so you have to move back to previous then 1024-896+452=580 usteps then you have to move back (-580*usteptommfactor). I'm just thinking how it should works, not read any document
```
effectorBackoutDir = -X_HOME_DIR;
stepperBackoutDir = INVERT_X_DIR ? effectorBackoutDir : -effectorBackoutDir;
```
effectorBackoutDir =-1 // reducing axis mm position moves away from end stop
stepperBackoutDir = 1 // increasing tmc usteps position moves away from endstop
896-452 = +444 tmc usteps
Increasing tmc usteps in the backout direction so it does not have to be inverted.
The number of tmc usteps to take is >0 so it doesn't have to skip 1024 more.

444 is the ammounto of tmc usteps we have to take, that makes 0.13736mm in the effectorBackoutDir (-1)

clear as mud 😄

**Comment by GMagician** [comment]
> stepperBackoutDir = 1 // increasing tmc usteps position moves away from endstop

Yes, this is right, my assumption was wrong on this...I though decreasing ustep moved away from endstop but since dir=0 going negative direction increments them...double inversion.

**Comment by GMagician** [comment]
@Fabi0San thanks for your support. I think now this is ready to merge

**Comment by Fabi0San** [comment]
> > stepperBackoutDir = 1 // increasing tmc usteps position moves away from endstop
> 
> Yes, this is right, my assumption was wrong on this...I though decreasing ustep moved away from endstop but since dir=0 going negative direction increments them...double inversion.

Yeah, all this inversions make my head hurt, it was worse before, this is the least number of inversions I could find.

>@Fabi0San thanks for your support. I think now this is ready to merge

Grazie Giuliano for finding the issue, creating the bug, wrapping your head around it and help coding the fix. 
I really appreciate it. 🍻 

@thinkyhead, we may be ready to merge now.

**Comment by GMagician** [comment]
@Fabi0San I wrote on Discord but none answered. 
_I'm wondering if the message "Selected home phase nnn too close to endstop trigger phase nnn. Pick a different phase for ?" shouldn't be echoed always and not only when debug is on. This may hide the cause of a variable home position and on Z may be problematic_
I think that once distance from one phase to other is at least twice than "tolerance for display message" this message may be echoed to alert users

**Comment by GMagician** [comment]
How fast...but it's what if axis resolution is too small? it may be possible that from phase to phase is less or near 0.05?

**Pull request merged**